presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/TestPushTopNThroughOuterJoin.java

```diff
@@ -0,0 +1,151 @@
+    public void testPushTopNThroughLeftJoin()
+    {
+        tester().assertThat(new PushTopNThroughOuterJoin())
+                .on(p -> {
+                    Symbol leftKey = p.symbol("leftKey");
+                    Symbol rightKey = p.symbol("rightKey");
+                    return p.topN(
+                            1,
+                            ImmutableList.of(leftKey),
+                            p.project(Assignments.of(leftKey, new SymbolReference("leftKey"), rightKey, new SymbolReference("rightKey")),
+                                    p.join(
+                                            LEFT,
+                                            p.values(5, leftKey),
+                                            p.values(5, rightKey),
+                                            new JoinNode.EquiJoinClause(leftKey, rightKey))));
+                })
+                .matches(
+                        topN(
+                                1,
+                                ImmutableList.of(sort("leftKey", ASCENDING, FIRST)),
+                                project(
+                                        join(
+                                                LEFT,
+                                                ImmutableList.of(equiJoinClause("leftKey", "rightKey")),
+                                                topN(1, ImmutableList.of(sort("leftKey", ASCENDING, FIRST)), Step.PARTIAL, values("leftKey")),
+                                                values("rightKey")))));
+    }
+
+    /**
+     * Push topN to right source of full outer join
+     */
+    @Test
+    public void testPushRightwardsTopNThroughFullOuterJoin()
+    {
+        tester().assertThat(new PushTopNThroughOuterJoin())
+                .on(p -> {
+                    Symbol leftKey = p.symbol("leftKey");
+                    Symbol rightKey = p.symbol("rightKey");
+                    return p.topN(
+                            1,
+                            ImmutableList.of(rightKey),
+                            p.project(Assignments.of(leftKey, new SymbolReference("leftKey"), rightKey, new SymbolReference("rightKey")),
+                                    p.join(
+                                            FULL,
+                                            p.values(5, leftKey),
+                                            p.values(5, rightKey),
+                                            new JoinNode.EquiJoinClause(leftKey, rightKey))));
+                })
+                .matches(
+                        topN(
+                                1,
+                                ImmutableList.of(sort("rightKey", ASCENDING, FIRST)),
+                                project(
+                                        join(
+                                                FULL,
+                                                ImmutableList.of(equiJoinClause("leftKey", "rightKey")),
+                                                values("leftKey"),
```

Comment:
why `topN` was not pushed here?

Reply:
TopN couldn't have been applied to the left source of the Join, as all OrderBy symbols (in this case the single symbol "rightKey") came from the right source.